Sources/MacOSPlatform/MacOS.swift

```diff
@@ -0,0 +1,184 @@
+public struct MacOS: Platform {
+    public init() {}
+
+    public var appDataDirectory: URL {
+        return FileManager.default.homeDirectoryForCurrentUser
+            .appendingPathComponent("Library/Application Support", isDirectory: true)
+    }
+
+    public var toolchainFileExtension: String {
+        "pkg"
+    }
+
+    public func isSystemDependencyPresent(_: SystemDependency) -> Bool {
+        // All system dependencies on macOS should be present
+        true
+    }
+
+    public func verifySystemPrerequisitesForInstall(requireSignatureValidation: Bool) throws {
+        // All system prerequisites should be there for macOS
+    }
+
+    public func install(from tmpFile: URL, version: ToolchainVersion) throws {
+        guard tmpFile.fileExists() else {
+            throw Error(message: "\(tmpFile) doesn't exist")
+        }
+
+        if !self.swiftlyToolchainsDir.fileExists() {
+            try FileManager.default.createDirectory(at: self.swiftlyToolchainsDir, withIntermediateDirectories: false)
+        }
+
+        SwiftlyCore.print("Installing package in user home directory...")
+        try runProgram("installer", "-pkg", tmpFile.path, "-target", "CurrentUserHomeDirectory")
+    }
+
+    public func uninstall(_ toolchain: ToolchainVersion) throws {
+        SwiftlyCore.print("Uninstalling package in user home directory...")
+
+        let toolchainDir = self.swiftlyToolchainsDir.appendingPathComponent("\(toolchain.identifier).xctoolchain", isDirectory: true)
+
+        let decoder = PropertyListDecoder()
+        let infoPlist = toolchainDir.appendingPathComponent("Info.plist")
+        guard let data = try? Data(contentsOf: infoPlist) else {
+            throw Error(message: "could not open \(infoPlist)")
+        }
+
+        guard let pkgInfo = try? decoder.decode(SwiftPkgInfo.self, from: data) else {
+            throw Error(message: "could not decode plist at \(infoPlist)")
+        }
+
+        try FileManager.default.removeItem(at: toolchainDir)
+
+        let homedir = ProcessInfo.processInfo.environment["HOME"]!
+        try runProgram("pkgutil", "--volume", homedir, "--forget", pkgInfo.CFBundleIdentifier)
+    }
+
+    public func use(_ toolchain: ToolchainVersion, currentToolchain: ToolchainVersion?) throws -> Bool {
+        let toolchainBinURL = self.swiftlyToolchainsDir
+            .appendingPathComponent(toolchain.identifier + ".xctoolchain", isDirectory: true)
+            .appendingPathComponent("usr", isDirectory: true)
+            .appendingPathComponent("bin", isDirectory: true)
+
+        // Delete existing symlinks from previously in-use toolchain.
+        if let currentToolchain {
+            try self.unUse(currentToolchain: currentToolchain)
+        }
+
+        // Ensure swiftly doesn't overwrite any existing executables without getting confirmation first.
+        let swiftlyBinDirContents = try FileManager.default.contentsOfDirectory(atPath: self.swiftlyBinDir.path)
```

Comment:
I wonder if it's worth us migration to NIOFileSystem, since we already have NIO as a dependency. Aside from the nicer API, it would be one less thing to use when depending on new Foundation which should make for a smaller binary

Reply:
I can raise this as a separate issue. I believe that there are Foundation dependencies all over the code. Pulling all of that out would be a significant sweep.

I'm not familiar with NIOFileSystem. Something that I really like about FileManager API is that it appears to be very mockable so that we could refactor the tests to operate on a full mock of the filesystem. Is that capability available with NIOFS?

Reply:
To be clear I'm not suggesting we pull out all of Foundation, but minimising it to just `FoundationEssentials` etc could help reduce the size (especially if we don't need ICU).

Which specific parts would you use for mocking (not getting into a discussion about that word 🤣 )? NIOFileSystem has a _similar_ API in that you have `FileSystem.shared` you perform the operations on, but the API is much more modern and built with Concurrency in mind (may also help solve some of the issues in #124 )

Reply:
In general, I agree that NIOFileSystem is more appropriate here since it uses async code to do the work. However, I would defer that for this PR since it is something we can do separately. WDYT @0xTim ?

Reply:
Yes let's defer this to another PR. I've added an issue #125

Reply:
Yeah absolutely shouldn't be a blocker for this PR